### PR TITLE
update contribution guidelines and add NEWS.md

### PR DESCRIPTION
**Pull request by bkamins**
Following the discussions on Slack I have made an update to contribution guidelines.

### CODE REVIEW

**Comment by oxinabox** [comment]
Should it have am introduction saying something like:
"We follow the ColPrac guide for collaborative practices. New contributor should make sure to read that guide.
Following here are some additional clarifications and practices we follow."

Do you think?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,19 +1,36 @@
+[![ColPrac: Contributor's Guide on Collaborative Practices for Community Packages](https://img.shields.io/badge/ColPrac-Contributor's%20Guide-blueviolet)](https://github.com/SciML/ColPrac)
```
Comment (bkamins) [inline CONTRIBUTING.md:1]:
```suggestion
[![ColPrac: Contributor's Guide on Collaborative Practices for Community Packages](https://img.shields.io/badge/ColPrac-Contributor's%20Guide-blueviolet)](https://github.com/SciML/ColPrac)

We follow the ColPrac guide for collaborative practices. New contributor should make sure to read that guide.
Following here are some additional clarifications and practices we follow.
```

**Comment by bkamins** [comment]
Sure, added. Thank you for looking into this.

Also - maybe you can consider putting a similar sentence in ColPrac when you suggest to add a badge?
Actually - I used only badge, because I concluded that you recommend in ColPrac to only put a badge :smile:.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,19 +1,39 @@
...
+  Increasing test coverage as you go is always nice
+* Aim for atomic commits, if possible, e.g. `change 'foo' behavior like so` &
+  `'bar' handles such and such corner case`,
+  rather than `update 'foo' and 'bar'` & `fix typo` & `fix 'bar' better`
+* Pull requests are tested against release and development branches of Julia,
+  so using `Pkg.test("DataFrames")` as you develop can be helpful
+* The style guidelines outlined below are not the personal style of most contributors,
+  but for consistency throughout the project, we've adopted them
 * It is recommended to disable GitHub Actions on your fork; check Settings > Actions
+* A PR with breaking changes should have `[BREAKING]` as a first part of its name
```
Comment (bkamins) [inline CONTRIBUTING.md:32]:
```suggestion
* If a PR adds a new exported name then make sure to add a docstring for it and
  add a reference to it in the documentation
* A PR with breaking changes should have `[BREAKING]` as a first part of its name
```

File: NEWS.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+* add new functions to `GroupKey` API to make it more consistent with `DataFrameRow`
+  ([#2308](https://github.com/JuliaData/DataFrames.jl/pull/2308))
+* allow column renaming in joins
+  (NOT MERGED YET [#2313](https://github.com/JuliaData/DataFrames.jl/pull/2313)
+
+## Deprecated
+
+* `DataFrame!` is now deprecated ([#2338](https://github.com/JuliaData/DataFrames.jl/pull/2338))
+* all old deprecations now throw an error
+  (NOT MERGED YET [#2350](https://github.com/JuliaData/DataFrames.jl/pull/2350))
```
Comment (bkamins) [inline NEWS.md:36]:
```suggestion
  ([#2350](https://github.com/JuliaData/DataFrames.jl/pull/2350))
```

File: NEWS.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+* add `filter` to `GroupedDataFrame` ([#2279](https://github.com/JuliaData/DataFrames.jl/pull/2279))
+* add `empty` and `empty!` function for `DataFrame` that remove all rows from it,
+  but keep columns ([#2262](https://github.com/JuliaData/DataFrames.jl/pull/2262))
+* make `indicator` keyword argument in joins allow passing a string
+  ([#2284](https://github.com/JuliaData/DataFrames.jl/pull/2284),
+   [#2296](https://github.com/JuliaData/DataFrames.jl/pull/2296))
+* add new functions to `GroupKey` API to make it more consistent with `DataFrameRow`
+  ([#2308](https://github.com/JuliaData/DataFrames.jl/pull/2308))
+* allow column renaming in joins
+  (NOT MERGED YET [#2313](https://github.com/JuliaData/DataFrames.jl/pull/2313)
```
Comment (bkamins) [inline NEWS.md:30]:
```suggestion
  ([#2313](https://github.com/JuliaData/DataFrames.jl/pull/2313)
```

**Comment by bkamins** [comment]
@nalimilan - can you please review? Thank you!

I would like to merge it and make sure that NEWS.md updates are made in new PRs that get submitted/merged (and if we wait too long with merging this one the state of NEWS.md it has now might get outdated).

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,19 +1,41 @@
...
+  so using `Pkg.test("DataFrames")` as you develop can be helpful
+* The style guidelines outlined below are not the personal style of most contributors,
+  but for consistency throughout the project, we've adopted them
 * It is recommended to disable GitHub Actions on your fork; check Settings > Actions
+* If a PR adds a new exported name then make sure to add a docstring for it and
+  add a reference to it in the documentation
+* A PR with breaking changes should have `[BREAKING]` as a first part of its name
+* If a PR changes or adds functionality please update NEWS.md file accordingly as
+  a part of the PR (along with the link to the PR); please do not add entries
+  to NEWS.md for changes that are bug fixes or are not user visible, such as
```
Comment (nalimilan) [inline CONTRIBUTING.md:37]: Wouldn't it be worth mentioning bug fixes in the news?

**Comment by nalimilan** [comment]
Maintaining NEWS.md is going to take some work, but that's up to you. :-p (I would find it easier to use the list that's generated automatically on the GitHub releases page, editing it manually if needed.)

**Comment by bkamins** [comment]
Editing GitHub releases page is always required unfortunately, as there are some problems with it every time.

I do not want to include bug fixes in the news exactly to make the effort of maintaining NEWS.md lower. Of course if we feel a bugfix is significant we can have it there but I do not want to commit to do it.

I add NEWS.md because this is what people want to have because reading GitHub releases page is not clear for them.

File: NEWS.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  ([#2291](https://github.com/JuliaData/DataFrames.jl/pull/2291))
+* `nunique` is not reported now by `describe` by default
+  ([#2339](https://github.com/JuliaData/DataFrames.jl/pull/2339))
+* stop reordering columns of the parent in `transform` and `transform!`;
+  always generate columns that were specified to be computed even for
+  `GroupedDataFrame` with zero rows
+  ([#2324](https://github.com/JuliaData/DataFrames.jl/pull/2324))
+* improve the rule for automatically generated column names in
+  `combine`/`select(!)`/`transform(!)` with composed functions
+  ([#2274](https://github.com/JuliaData/DataFrames.jl/pull/2274))
```
Comment (bkamins) [inline NEWS.md:17]:
```suggestion
  ([#2274](https://github.com/JuliaData/DataFrames.jl/pull/2274))
* `:nmissing` in `describe` now produces `0` if the column does not allow
   missing values; earlier `nothing` was produced in this case
   ([#2360](https://github.com/JuliaData/DataFrames.jl/pull/2360))
```

File: NEWS.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+* `:nmissing` in `describe` now produces `0` if the column does not allow
+   missing values; earlier `nothing` was produced in this case
+   ([#2360](https://github.com/JuliaData/DataFrames.jl/pull/2360))
```
Comment (bkamins) [inline NEWS.md:20]:
```suggestion
   ([#2360](https://github.com/JuliaData/DataFrames.jl/pull/2360))
* fast aggregation functions in for `GroupedDataFrame` now correctly
   choose the fast path only when it is safe; this resolves inconsistencies
   with what the same functions not using fast path produce
   ([#2357](https://github.com/JuliaData/DataFrames.jl/pull/2357))
```

File: NEWS.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  ([#2339](https://github.com/JuliaData/DataFrames.jl/pull/2339))
+* stop reordering columns of the parent in `transform` and `transform!`;
+  always generate columns that were specified to be computed even for
+  `GroupedDataFrame` with zero rows
+  ([#2324](https://github.com/JuliaData/DataFrames.jl/pull/2324))
+* improve the rule for automatically generated column names in
+  `combine`/`select(!)`/`transform(!)` with composed functions
+  ([#2274](https://github.com/JuliaData/DataFrames.jl/pull/2274))
+* `:nmissing` in `describe` now produces `0` if the column does not allow
+   missing values; earlier `nothing` was produced in this case
```
Comment (bkamins) [inline NEWS.md:19]:
```suggestion
  missing values; earlier `nothing` was produced in this case
```

File: NEWS.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+* add `filter` to `GroupedDataFrame` ([#2279](https://github.com/JuliaData/DataFrames.jl/pull/2279))
+* add `empty` and `empty!` function for `DataFrame` that remove all rows from it,
+  but keep columns ([#2262](https://github.com/JuliaData/DataFrames.jl/pull/2262))
+* make `indicator` keyword argument in joins allow passing a string
+  ([#2284](https://github.com/JuliaData/DataFrames.jl/pull/2284),
+   [#2296](https://github.com/JuliaData/DataFrames.jl/pull/2296))
+* add new functions to `GroupKey` API to make it more consistent with `DataFrameRow`
+  ([#2308](https://github.com/JuliaData/DataFrames.jl/pull/2308))
+* allow column renaming in joins
+  ([#2313](https://github.com/JuliaData/DataFrames.jl/pull/2313)
```
Comment (bkamins) [inline NEWS.md:37]:
```suggestion
  ([#2313](https://github.com/JuliaData/DataFrames.jl/pull/2313)
 * add `rownumber` to `DataFrameRow` ([#2356](https://github.com/JuliaData/DataFrames.jl/pull/2356))
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,19 +1,41 @@
...
 ## Reporting Issues
 
-* It's always good to start with a quick search for an existing issue to post on, or related issues for context, before opening a new issue
-* Including minimal examples is greatly appeciated
-* If it's a bug, or unexpected behaviour, reproducing on the latest development version (`Pkg.checkout("DataFrames")`) is a good gut check and can streamline the process, along with including the first two lines of output from `versioninfo()`
+* It's always good to start with a quick search for an existing issue to post on,
+  or related issues for context, before opening a new issue
+* Including minimal examples is greatly appreciated
+* If it's a bug, or unexpected behaviour, reproducing on the latest development version
+  (`Pkg.checkout("DataFrames")`) is a good gut check and can streamline the process,
```
Comment (KristofferC) [inline CONTRIBUTING.md:14]: This doesn't exist anymore

Comment (bkamins) [reply CONTRIBUTING.md:14]: Thank you!

Comment (KristofferC) [reply CONTRIBUTING.md:14]: `Pkg.add(name="DataFrames", branch="master")` maybe is a good replacement.

Comment (bkamins) [reply CONTRIBUTING.md:14]: ah - in parallel I have opened a PR with `Pkg.develop` recommendation. Can you please answer there what you prefer. Thank you!
